File: juno/indicators/dm.py
from decimal import Decimal
from typing import Tuple


# Directional Movement Indicator
class DM:
    plus_value: Decimal = Decimal('0.0')
    minus_value: Decimal = Decimal('0.0')

    _per: Decimal

    _dmup: Decimal = Decimal('0.0')
    _dmdown: Decimal = Decimal('0.0')
    _prev_high: Decimal = Decimal('0.0')
    _prev_low: Decimal = Decimal('0.0')

    _t: int = 0
    _t1: int = 2
    _t2: int
    _t3: int

    def __init__(self, period: int) -> None:
        if period < 1:
            raise ValueError(f'Invalid period ({period})')

        self._per = (period - 1) / Decimal(period)

        self._t2 = period
        self._t3 = period + 1

    @property
    def maturity(self) -> int:
        return self._t2

    @property
    def mature(self) -> bool:
        return self._t >= self._t2

    @property
    def diff(self) -> Decimal:
        return abs(self.plus_value - self.minus_value)

    @property
    def sum(self) -> Decimal:
        return self.plus_value + self.minus_value

    def update(self, high: Decimal, low: Decimal) -> Tuple[Decimal, Decimal]:
        self._t = min(self._t + 1, self._t3)

        if self._t >= self._t1 and self._t < self._t3:
            dp, dm = _calc_direction(self._prev_high, self._prev_low, high, low)
            self._dmup += dp
            self._dmdown += dm

        if self._t == self._t2:
            self.plus_value = self._dmup
            self.minus_value = self._dmdown
        elif self._t >= self._t3:
            dp, dm = _calc_direction(self._prev_high, self._prev_low, high, low)
            self._dmup = self._dmup * self._per + dp
            self._dmdown = self._dmdown * self._per + dm
            self.plus_value = self._dmup
            self.minus_value = self._dmdown

        self._prev_high = high
        self._prev_low = low
        return self.plus_value, self.minus_value


def _calc_direction(prev_high: Decimal, prev_low: Decimal, high: Decimal,
                    low: Decimal) -> Tuple[Decimal, Decimal]:
    up = high - prev_high
    down = prev_low - low

    if up < 0:
        up = Decimal('0.0')
    elif up > down:
        down = Decimal('0.0')

    if down < 0:
        down = Decimal('0.0')
    elif down > up:
        up = Decimal('0.0')

    return up, down

File: juno_rs/src/indicators/dm.rs
use std::cmp::min;

pub struct DM {
    pub plus_value: f64,
    pub minus_value: f64,
    per: f64,
    dmup: f64,
    dmdown: f64,
    prev_high: f64,
    prev_low: f64,
    t: u32,
    t1: u32,
    t2: u32,
    t3: u32,
}

impl DM {
    pub fn new(period: u32) -> Self {
        Self {
            plus_value: 0.0,
            minus_value: 0.0,
            dmup: 0.0,
            dmdown: 0.0,
            per: f64::from(period - 1) / f64::from(period),
            prev_high: 0.0,
            prev_low: 0.0,
            t: 0,
            t1: 1,
            t2: period - 1,
            t3: period,
        }
    }

    pub fn maturity(&self) -> u32 {
        self.t2
    }

    pub fn mature(&self) -> bool {
        self.t >= self.t2
    }

    pub fn update(&mut self, high: f64, low: f64) {
        if self.t >= self.t1 && self.t < self.t2 {
            let (dp, dm) = calc_direction(self.prev_high, self.prev_low, high, low);
            self.dmup += dp;
            self.dmdown += dm;
        }

        if self.t == self.t2 {
            self.plus_value = self.dmup;
            self.minus_value = self.dmdown;
        } else if self.t == self.t3 {
            let (dp, dm) = calc_direction(self.prev_high, self.prev_low, high, low);
            self.dmup = self.dmup * self.per + dp;
            self.dmdown = self.dmdown * self.per + dm;
            self.plus_value = self.dmup;
            self.minus_value = self.dmdown;
        }

        self.prev_high = high;
        self.prev_low = low;
        self.t = min(self.t + 1, self.t3);
    }
}

fn calc_direction(prev_high: f64, prev_low: f64, high: f64, low: f64) -> (f64, f64) {
    let mut up = high - prev_high;
    let mut down = prev_low - low;

    if up < 0.0 {
        up = 0.0;
    } else if up > down {
        down = 0.0;
    }

    if down < 0.0 {
        down = 0.0;
    } else if down > up {
        up = 0.0;
    }

    (up, down)
}
